### PR TITLE
chore(deps): use nanoarrow in the snowflake backend

### PR DESCRIPTION
**Pull request by cpcloud**
This PR has been a long time coming :) We are finally able to support multiple versions of pyarrow--including the latest version--thanks to the efforts of the folks at Snowflake and Voltron Data and their work on nanoarrow. Notably, the Snowflake test suite is passing with no changes to ibis other than a dependency bump.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -87,7 +87,7 @@ shapely = { version = ">=2,<3", optional = true }
 # also, we don't support arbitrarily old versions of this library
 snowflake-connector-python = { version = ">=3.0.2,<4", optional = true, extras = [
   "pandas",
-] }
```
Comment (cpcloud) [inline pyproject.toml:90]: I plan to remove this once the release is out. I'll move this out of draft mode then as well.

File: pyproject.toml [hunk 2]
```diff
@@ -87,7 +87,7 @@ shapely = { version = ">=2,<3", optional = true }
...
+], allow-prereleases = true }
```
Comment (cpcloud) [inline pyproject.toml:90]: I plan to remove this once the release is out. I'll move this out of draft mode then as well.

**Comment by cpcloud** [comment]
Snowflake backend tests lookin' good:

```
…/ibis on  snowflake-nanoarrow is 📦 v6.0.0 via 🐍 v3.10.12 via ❄️  impure (ibis-3.10.12-env) took 5s
❯ pytest -m snowflake -n auto --dist loadgroup -q
bringing up nodes...
.............................xx..x......x........x.....x.........................x............................x..........x....x.....x...............x............xx..xxx...................... [ 19%]
...........x...................................x............xx.x..x...............x.....x..xx.........x.................................x..x..............x........x.........x.x.....x...x...x [ 38%]
..............xx.......xxx.x...xx.x.............x.................................x......x.......................................xx......x............xxx.....x............x..x.....x.....xxxx [ 57%]
x..x.x......x..x..x..............x.xxx.x..xxxxxxxxxxxx........x.x...x.........x....x..x.xx..x.....x.....x.......xx.....x.....x..............x.....xx........x......x...xx..xx.....xx...x...... [ 76%]
...x..........s.............x......x..x..........x.x...x..x...xx.......s...x....x.........x.....x..x.xx.......x..x..s........xx.s....xx.....x..s..........x.......x...s.....x..x.....x.x..xxx. [ 96%]
..xxxx.s.x..xxxx.xxx...x..xxxx..xxxxxx                                                                                                                                                         [100%]
812 passed, 7 skipped, 169 xfailed in 288.59s (0:04:48)
```

**Comment by cpcloud** [comment]
Latest version (`3.1.0a2`) of the alpha release is passing:

```
…/ibis on  snowflake-nanoarrow is 📦 v6.1.0 via 🐍 v3.10.12 via ❄️  impure (ibis-3.10.12-env) took 5s
❯ pytest -m snowflake -n auto --dist loadgroup -q
bringing up nodes...
......x........x.x...........xxx..x........x..............x...........x........x...........xx.......x...xxxxxx.....xx...x....xs............x..........x.....................x..........x...... [ 16%]
.........x..............x.......x....x.....x..x.xx.....................x............x.....x................................x........................x...............x...x..........x.x........ [ 32%]
......x...xx....x..........................xx......x................x...x..........xxx...x............x...x......x.xx.x....x....x..x.................x.xx.xx..........s.xx.s...x....s......... [ 48%]
........x......................x...........x....x....x...........x.x...x...................x........x..x.........................xx.......................................x......x...x........ [ 64%]
.x...................................x.......x....x.............x.....................x.x...........xxx.x.xxxxxxxxxx.x.xxxxxxxxxxx.x.................x.x.........x............................ [ 81%]
........................................s...s..................xx...x...............................x.................................................x....xx....xxx.....x.....xx.x..x......xx [ 97%]
xxx..x.x.x.x.xxxxx.xxxxxxxxxxx                                                                                                                                                                 [100%]
996 passed, 6 skipped, 168 xfailed in 288.38s (0:04:48)
```

**Review by gforsyth — APPROVED** [review]
This looks good to go, pending the actual release of whatever version of the connector ships the `nanoarrow` changes.

**Comment by cpcloud** [comment]
I'm going to pull out the PyArrow 12 related changes here into a separate PR, since those can go in now.

**Comment by cpcloud** [comment]
This PR is just adding visual noise at the moment. nanoarrow support is still in beta (finally out of alpha?).

Realistically we're going to cut another release before an official `snowflake-connector-python` release with nanoarrow is out.

Longer term, we are going to investigate using the snowflake adbc connector, and at the very least move the snowflake backend to sqlglot to avoid a dependency on `snowflake-sqlalchemy` which is one of the things blocking use of ibis inside snowpark.

**Pull request closed without merging**